
package protocol.Baidu;

//protocol version 2.3

message BidRequest {
	//  请求 ID,  唯一标识本次请求，明文字符串
	required string id = 1;
	// ****  用户信息  ****
	//  用户 IP 地址，点分十进制字符串
	optional string ip = 2;
	// User-Agent
	optional string user_agent = 3;
	//  用户 ID，该将于 2015 年 4 月 30 日停止使用
	optional string baidu_user_id = 4;
	//  用户 ID 版本，该将于 2015 年 4 月 30 日停止使用
	optional int32 baidu_user_id_version = 5;
	//  用户 ID
	//  在 2015 年 4 月 30 日之前将会同时发送版本 1 和版本 2 的 cookie
	//  从 2015 年 5 月 1 日起只发送版本 2 的 cookie
	message BaiduId {
	optional string baidu_user_id = 1;
	optional int32 baidu_user_id_version = 2;
	}
	repeated BaiduId baidu_id_list = 8;
	//  用户兴趣
	repeated int64 user_category = 6;
	//  用户性别
	enum Gender {
	UNKNOWN = 0;
	MALE = 1;
	FEMALE = 2;
	}
	optional Gender gender = 7;
	//  页面语言
	optional string detected_language = 9;
	// flash 版本
	optional string flash_version = 10;
	// ****  位置信息  ****
	message Geo {
	//  经纬度信息
	message Coordinate {
	//  经纬度坐标标准
	enum Standard {
	//  百度地图的经纬度坐标标准
	BD_09 = 0;
	//  国测局制定的经纬度坐标标准
	GCJ_02 = 1;
	//  国际经纬度坐标标准
	WGS_84 = 2;
	}
	//  地图坐标标准
	optional Standard standard = 1;
	//  纬度
	optional float latitude = 2;
	//  经度
	optional float longitude = 3;
	}
	repeated Coordinate user_coordinate = 1;
	//  用户位置信息
	message UserLocation {
	//  省份，如北京市
	optional string province = 1;
	//  城市，如北京市
	optional string city = 2;
	//  区县，如海淀区
	optional string district = 3;
	//  街道，如中关村大街
	optional string street = 4;
	}
	optional UserLocation user_location = 2;
	}
	optional Geo user_geo_info = 28;
	// ****  页面信息  ****
	//  当前页面 URL
	optional string url = 11;
	//  请求的 referer
	optional string referer = 12;
	//  网站分类
	optional int32 site_category = 13;
	//  网站质量类型
	optional int32 site_quality = 14;
	//  页面类型
	optional int32 page_type = 15;
	//  页面关键词
	repeated string page_keyword = 17;
	//  页面内容质量
	optional int32 page_quality = 18;
	//  页面分类
	optional int32 page_vertical = 21;
	//  发布商不允许的广告行业
	repeated int32 excluded_product_category = 19 [packed=true];
	// ****  移动信息  ****
	message Mobile {
	// ****  移动设备信息  ****
	//  百度唯一标识移动设备的编号
	optional string device_id = 1;
	//  设备类型定义
	enum MobileDeviceType {
	UNKNOWN_DEVICE = 0;
	HIGHEND_PHONE = 1;
	TABLET = 2;
	}
	//  设备类型
	optional MobileDeviceType device_type = 2;
	//  移动平台名，例如 android，iphone 等等
	enum OS {
	UNKNOWN_OS = 0;
	IOS = 1;
	ANDROID = 2;
	WINDOWS_PHONE = 3;
	}
	optional OS platform = 3 [default = UNKNOWN_OS];
	//  移动操作系统版本号
	//  例如  Android 2.1, major, minor 分别是 2,1
	//  例如  Iphone 4.2.1，major, minor, micro 分别是 4,2,1
	message DeviceOsVersion {
	optional int32 os_version_major = 1;
	optional int32 os_version_minor = 2;
	optional int32 os_version_micro = 3;
	}
	optional DeviceOsVersion os_version = 4;
	//  设备品牌
	optional string brand = 5;
	//  设备机型
	optional string model = 6;
	//  设备屏宽
	optional int32 screen_width = 7;
	//  设备屏高
	optional int32 screen_height = 8;
	//  运营商
	optional int64 carrier_id = 9;
	//  无线网络类型
	enum WirelessNetworkType {
	UNKNOWN_NETWORK = 0;
	WIFI = 1;
	MOBILE_2G = 2;
	MOBILE_3G = 3;
	MOBILE_4G = 4;
	}
	optional WirelessNetworkType wireless_network_type = 10;
	//  移动设备上为广告控制提供的 ID
	optional string for_advertising_id = 11;
	// ****  移动应用信息  ****
	message MobileApp {
	//  百度移动联盟为该 App 分配的 app id
	optional string app_id = 1;
	// App 应用的 package 全名
	optional string app_bundle_id = 2;
	// App 应用分类
	optional int32 app_category = 3;
	// App 开发者 ID
	optional int32 app_publisher_id = 4;
	// App 允许的交互类型定义
	//  电话、下载
	enum AppInteractionType {
	TELEPHONE = 0;
	DOWNLOAD = 1;
	}
	// App 允许的交互类型
	repeated AppInteractionType app_interaction_type = 5;
	}
	//  移动应用信息，当流量来自 App 时该字段非空
	optional MobileApp mobile_app = 12;
	}
	//  移动信息，当流量来自移动设备时该字段非空
	optional Mobile mobile = 29;
	// ****  视频信息  ****
	message Video {
	//视频的标题
	optional string title = 1;
	//视频的标签
	repeated string tags = 2;
	//视频的播放时长
	optional int32 content_length = 3;
	//视频的频道 ID
	repeated int64 channel_id = 4;
	}
	optional Video video = 30;
	// ****  广告位信息  ****
	message AdSlot {
	//  广告位 ID，全局唯一 ID
	optional uint64 ad_block_key = 1;
	//  当前页面广告位顺序 id，同一页面从 1 开始
	optional int32 sequence_id = 2;
	//  展示类型
	optional int32 adslot_type = 3;
	//  宽
	optional int32 width = 4;
	//  高
	optional int32 height = 5;
	//  视频广告的最大时长
	optional int32 max_video_duration = 10;
	//  视频广告的最小时长
	optional int32 min_video_duration = 11;
	//  视频广告的播放延时，  0 及正整数表示前贴，-1 表示中贴，-2 表示后贴
	optional int32 video_start_delay = 12;
	//  展示位置
	optional int32 slot_visibility = 6;
	//  发布商允许的创意类型
	repeated int32 creative_type = 7 [packed=true];
	//  发布商不允许的 landing page url
	repeated string excluded_landing_page_url = 8;
	//  媒体保护设置对应信息的 ID
	repeated fixed64 publisher_settings_list_id = 14;
	//  发布商设置的底价，单位分
	optional int32 minimum_cpm = 9;
	//  优先交易信息
	message PreferredOrderInfo {
	message PreferredOrder {
	//  优先交易 ID
	optional int64 order_id = 1;
	//  该优先交易的交易价格
	optional int64 fixed_cpm = 2;
	}
	//  一次请求可能包含多个优先交易
	repeated PreferredOrder preferred_orders = 1;
	//  允许优先交易的同时，是否允许竞价
	optional bool allow_auction = 2 [default=true];
	}
	optional PreferredOrderInfo preferred_order_info = 13;
	}
	//  默认每次请求一个广告位
	repeated AdSlot adslot = 20;
	// ****  系统使用  ****
	optional bool is_test = 26 [default=false];
	optional bool is_ping = 27 [default=false];
}

message BidResponse {
	//  返回 ID，将请求中的 id 复制给返回 id，便于 session trace
	required string id = 1;
	// ****  竞价广告信息  ****
	message Ad {
	//  广告位顺序 ID
	optional int32 sequence_id = 1;
	//  创意 ID
	optional int64 creative_id = 2;
	//  最高竞价，单位分
	optional int32 max_cpm = 3;
	//  扩展参数
	optional string extdata = 5;
	//  是否进行 cookie matching
	optional bool is_cookie_matching = 6;
	//  代码段
	optional string html_snippet = 7;
	//  广告主 id
	optional uint32 advertiser_id = 8;
	//  创意宽
	optional int32 width = 9;
	//  创意高
	optional int32 height = 10;
	//  创意所属行业
	optional int32 category = 11;
	//  创意类型
	optional int32 type = 12;
	//  到达页面
	optional string landing_page = 13;
	//  点击链接
	repeated string target_url = 14;
	//  如果选择优先交易，该交易的 ID
	optional int64 preferred_order_id = 15;
	}
	repeated Ad ad = 2;
	// ****  系统使用  ****
	// debug 接口
	optional string debug_string = 3;
	// DSP 处理时间
	optional int32 processing_time_ms = 4;
}